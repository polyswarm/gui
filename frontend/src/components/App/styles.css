.App {
  background-color: #f5f5f5;
  background-image: url(gray_hexgrid_2x.png);
  background-size: 129px 149px;
  background-repeat: repeat;
  height: 100vh;
  width: 100vw;
}

.App-Contents {
  padding: 20px;
}
