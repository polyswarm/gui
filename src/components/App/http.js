import BigNumber from 'bignumber.js';
import validator from 'validator';
import web3Utils from 'web3-utils';

class HttpApp {
  constructor(url, ws) {
    this.url = url;
    this.ws = ws;
  }

  getUnlockedWallet() {
    return fetch(this.url+'/accounts/active')
      .then(response => {
        if (response.ok) {
          return response;
        } else {
          throw new Error('Unable to access accounts');
        }
      })
      .then(response => response.json())
      .then(json => json.result)
      .catch(() => null);
  }

  getWallets() {
    return fetch(this.url+'/accounts')
      .then(response => {
        if (response.ok) {
          return response;
        } else {
          throw new Error('Unable to access accounts');
        }
      })
      .then(response => response.json())
      .then(json => json.result)
      .catch(() => []);
  }

  getBounty(bounty) {
    return new Promise((resolve, reject) => {
      if (validator.isUUID(bounty.guid, 4)) {
        resolve(bounty.guid);
      } else {
        reject('Invalid GUID');
      }
    })
      .then((guid) => fetch(this.url+'/bounties/'+guid))
      .then(response => {
        if (response.ok) {
          return response;
        } else {
          throw Error('Cannot get bounties.');
        }
      })
      
      .then(response => response.json())
      .then(json => json.result)
      .then(bounty => this.getBountyIsActive(bounty))
      .then(bounty => {
        const amount = new BigNumber(bounty.amount).dividedBy(new BigNumber('1000000000000000000')).toNumber();
        bounty.amount = amount;
        bounty.type = 'bounty';
        return bounty;
      })
      .then(bounty => this.getAssertionsForBounty(bounty))
      .then(bountyAssertions => this.getArtifactsForBounty(bountyAssertions))
      .catch(() => null);
  }

  getOffer(offer) {
    return new Promise((resolve, reject) => {
      if (validator.isUUID(offer.guid, 4)) {
        resolve(offer.guid);
      } else {
        reject('Invalid GUID');
      }
    })
      .then(() => new Promise((resolve) => resolve(offer)))
      .catch(() => null);
  }

  getArtifactsForBounty(bounty) {
    return fetch(this.url+'/artifacts/'+bounty.uri)
      .then(response => {
        if (response.ok) {
          return response;
        } else {
          throw new Error('Unable to access accounts');
        }
      })
      .then(response => response.json())
      .then(json => json.result)
      .then(files => {
        return files.map((file) => {
          return file.name;
        });
      })
      .then(filesnames => {
        return filesnames.map((name) => {
          const trimmed = name.trim();
          return {name: trimmed, good: 0, total: 0, assertions: []};
        });
      })
      .then(files => {
        bounty.artifacts = files;
        return bounty;
      });
  }

  getAssertionsForBounty(bounty) {
    return new Promise((resolve, reject) => {
      if (validator.isUUID(bounty.guid, 4)) {
        resolve(bounty.guid);
      } else {
        reject('Invalid GUID');
      }
    })
      .then(guid => fetch(this.url+'/bounties/'+guid+'/assertions'))
      .then(response => {
        if (response.ok) {
          return response;
        } else {
          throw new Error('Unable to access accounts');
        }
      })
      .then(response => response.json())
      .then(json => json.result)
      .then(assertions => {
        return assertions.map((assertion) => {
          const bid = new BigNumber(assertion.bid).dividedBy(new BigNumber('1000000000000000000')).toNumber();
          return {
            author: assertion.author,
            bid: bid,
            verdicts: assertion.verdicts,
            metadata: assertion.metadata,
          };
        });
      })
      .then(filtered => {
        bounty.assertions = filtered;
        return bounty;
      });
  }

  getBountyIsActive(bounty) {
    return fetch(this.url + '/bounties/active')
      .then(response => {
        if (response.ok) {
          return response;
        } else {
          throw new Error('Unable to access active bounties');
        }
      })
      .then(response => response.json())
      .then(json => json.result)
      .then(bounties => bounties.findIndex(b => b.guid === bounty.guid) >= 0 )
      .then(found => {
        bounty.expired = !found;
        return bounty;
      });
  }

  getEth(wallet) {
    const url = this.url;
    return new Promise((resolve, reject) => {
      if (web3Utils.isAddress(wallet)) {
        resolve(wallet);
      } else {
        reject(`${wallet} is not an Ethereum address`);
      }
    })
      .then(address => fetch(url+'/accounts/'+address+'/balance/eth'))
      .then(response => {
        if (response.ok) {
          return response;
        }
        throw Error('Failed to get balance');
      })
      .then(response => response.json())
      .then(json => json.result+'')
      .catch(() => 0);
  }

  getNct(wallet) {
    const url = this.url;
    return new Promise((resolve, reject) => {
      if (web3Utils.isAddress(wallet)) {
        resolve(wallet);
      } else {
        reject(`${wallet} is not an Ethereum address`);
      }
    })
      .then(address => fetch(url+'/accounts/'+address+'/balance/nct'))
      .then(response => {
        if (response.ok) {
          return response;
        }
        throw Error('Failed to get balance');
      })
      .then(response => response.json())
      .then(json => json.result+'')
      .catch(() => 0);
  }

  listenForAssertions(assertionAddedCallback) {
    // attach to websocket
    // anytime we get an assertion, check if it matches a guid
    // if it does, add it to the assertions for that object
    const ws = this.ws;
    const websocket = new WebSocket(ws);

    websocket.onmessage = (event) => {
      const message = JSON.parse(event.data);

      if (message.event === 'assertion') {
        const body = message.data;
        const bid = new BigNumber(body.bid).dividedBy(new BigNumber('1000000000000000000')).toNumber();
        const assertion = {
          guid: body.bounty_guid,
          bid: bid,
          verdicts: body.verdicts,
          metadata: body.metadata,
          author: body.author,
        };
        assertionAddedCallback(assertion);
      }
    };
  }
}
export default HttpApp;
