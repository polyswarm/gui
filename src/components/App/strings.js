const strings = {
  createBounty: 'Create a Bounty',
  createOffer: 'Create an Offer',
  list: 'Offers & Bounties',
  bountyError: 'Error: Failed to Post Bounty',
  errorOffer: 'Error: Failed to Open Offer Channel',
  requestAllData: 'Refreshing',
  requestWallets: 'Updating Wallets',
  requestGetBounty: 'Getting Bounty',
  requestGetOffer: 'Getting Offer',
};

export default strings;
