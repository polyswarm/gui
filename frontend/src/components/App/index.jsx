import React, { Component } from 'react';
import Header from '../Header';
import AddressField from '../AddressField';
import ChainInfo from '../ChainInfo';
import TransferForm from '../TransferForm';
import Button from '../Button';
import './styles.css';

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      address: null,
    }

    this.onAddressChange = this.onAddressChange.bind(this);
  }
  render() {
    return (
      <div className="App">
        <Header />
        <div className="App-Contents">
          <AddressField onChange={this.onAddressChange} />
          <div className="App-Chains">
            <ChainInfo />
            <ChainInfo />
          </div>
          <TransferForm />
        </div>
      </div>
    );
  }

  onAddressChange(address) {
    this.setState({address: address});
  }
}

export default App;
