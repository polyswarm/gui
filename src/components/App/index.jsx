// Vendor imports
import React, { Component } from 'react';
import Uuid from 'uuid/v4';
import {CSSTransition} from 'react-transition-group';
import BigNumber from 'bignumber.js';
// Bounty imports
import BountyCreate from '../BountyCreate';
import BountyInfo from '../BountyInfo';
import BountyList from '../BountyList';
import OfferCreate from '../OfferCreate';
import OfferInfo from '../OfferInfo';
import Snackbar from '../Snackbar';
import Welcome from '../Welcome';
// Component imports
import HttpApp from './http';
import config from '../../config';
import strings from './strings';

class App extends Component {
  constructor(props) {
    super(props);
    const {bounties, first} = this.preloadLocalStorage();
    this.http = new HttpApp(config.host, config.websocket_host);
    this.state = {
      address: 0,
      walletList: [],
      active: -1,
      bounties: bounties,
      createBounty: false,
      createOffer: false,
      first: first,
      errorMessage: null,
      requestsInProgress: []
    };

    this.onAddBounty = this.onAddBounty.bind(this);
    this.onAddOffer = this.onAddOffer.bind(this);
    this.onBackPressed = this.onBackPressed.bind(this);
    this.onRemoveBounty = this.onRemoveBounty.bind(this);
    this.onSelectBounty = this.onSelectBounty.bind(this);
    this.onCreateBounty = this.onCreateBounty.bind(this);
    this.onCreateOffer = this.onCreateOffer.bind(this);
    this.onCloseWelcome = this.onCloseWelcome.bind(this);
    this.onErrorDismissed = this.onErrorDismissed.bind(this);
    this.onPostError = this.onPostError.bind(this);
    this.onWalletChangeHandler = this.onWalletChangeHandler.bind(this);
    this.addRequest = this.addRequest.bind(this);
    this.removeRequest = this.removeRequest.bind(this);
    this.getData = this.getData.bind(this);
    this.getWallets = this.getWallets.bind(this);
    this.updateOnAssertion = this.updateOnAssertion.bind(this);
  }

  componentDidUpdate(_, prevState) {
    const { state: { bounties } } = this;
    const { bounties: prevBounties } = prevState;
    const storageOutOfSync =  JSON.stringify(bounties) !== JSON.stringify(prevBounties);
    if(storageOutOfSync) {
      this.storeBounties(bounties);
    }
  }

  componentDidMount() {
    this.getData();
    this.getWallets();
    this.timer = setInterval(() => {
      this.getWallets();
    }, 5000);
  }

  componentWillUnmount() {
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = null;
    }
  }

  render() {

    const { state: { active, bounties, createBounty, createOffer, first, errorMessage } } = this;

    return (
      <div className='App'>
        <CSSTransition
          in={first}
          timeout={200}
          mountonEnter
          unmountOnExit
          classNames='fade'>
          {() => (
            <Welcome onClick={this.onCloseWelcome}/>
          )}
        </CSSTransition>
        {!first && (
          <React.Fragment>
            { createBounty && (
              <BountyCreate
                {...this.getPropsForChild()}
                onWalletChange={this.onWalletChangeHandler}
                addBounty={this.onAddBounty}
                onBountyPosted={this.onBackPressed}/>
            )}
            { createOffer && (
              <OfferCreate
                {...this.getPropsForChild()}
                onWalletChange={this.onWalletChangeHandler}
                addOffer={this.onAddOffer}
                onBountyPosted={this.onBackPressed}/>
            )}
            { !createBounty && !createOffer && active < 0 && (
              <BountyList
                {...this.getPropsForChild()}
                onBountySelected={this.onSelectBounty}
                onBountyRemoved={this.onRemoveBounty}/>
            )}
            { !createBounty && active >=0 && bounties[active].type === 'bounty' && (
              <BountyInfo
                {...this.getPropsForChild()}
                bounty={bounties[active]}/>
            )}
            { !createOffer && active >=0 && bounties[active].type === 'offer' && (
              <OfferInfo
                {...this.getPropsForChild()}
                offer={bounties[active]}/>
            )}
            {errorMessage && errorMessage.length > 0 && (
              <Snackbar message={errorMessage}
                onDismiss={this.onErrorDismissed}/>
            )}
          </React.Fragment>
        )}
      </div>
    );
  }

  onAddBounty(result) {
    const http = this.http;

    this.addRequest(strings.requestGetBounty, result.guid);
    return http.getBounty(result)
      .then(bounty => {
        if (bounty != null) {
          bounty.updated = true;
          const bounties = this.state.bounties.slice();
          bounties.push(bounty);
          this.setState({bounties: bounties});
        }
      })
      .catch(() => {})
      .then(() => {
        this.removeRequest(strings.requestGetBounty, result.guid);
      });
  }

  onAddOffer(result) {

  }

  onBackPressed() {
    this.setState({active: -1, createBounty: false, createOffer: false});
  }

  onCreateBounty() {
    this.setState({createBounty: true, active: -1});
  }

  onCreateOffer() {
    this.setState({createOffer: true, active: -1});
  }

  onCloseWelcome() {
    this.setState({first: false});
    this.markSeen();
  }

  onErrorDismissed() {
    this.setState({errorMessage: null});
  }

  onPostError(message) {
    this.setState({errorMessage: message});
  }

  onRemoveBounty(index) {
    const bounties = this.state.bounties.slice();
    const { state: {active}} = this;
    if (index !== null && index >= 0 && index < bounties.length) {
      bounties.splice(index, 1);
      let activeValue = active;
      if (active >= bounties.length) {
        activeValue = active - 1;
      }
      this.setState({active: activeValue, bounties: bounties});
    }
  }

  onSelectBounty(index) {
    /*
     * Need a deep copy to edit a specific value on an object in the array
     * without modifying state.
     *
     * FIXME: The stringify stuff is temporary. Will probably change how selected works
     * so that I can do a shallow copy of a different array.
     */
    const bounties = JSON.parse(JSON.stringify(this.state.bounties.slice()));
    if (index !== null && index >= 0 && index < bounties.length) {
      bounties[index].updated = false;
      this.setState({active: index, create: false, bounties: bounties});
    }
  }

  onWalletChangeHandler(store) {
    this.setState({isUnlocked: store});
    this.getWallets();
  }

  removeRequest(title, id/*, success*/) {
    const requestsInProgress = this.state.requestsInProgress.slice();
    const index = requestsInProgress.findIndex((obj) => obj.id == id);
    if (index >= 0 ) {
      requestsInProgress.splice(index, 1);
      this.setState({requestsInProgress: requestsInProgress});
      // TODO notify users of the success of the request
    }
  }

  addRequest(title, id) {
    const requestsInProgress = this.state.requestsInProgress.slice();
    requestsInProgress.push({title: title, id: id});
    this.setState({requestsInProgress: requestsInProgress});
  }

  updateOnAssertion(assertion) {
    const bounties = this.state.bounties.slice();
    const guid = assertion.guid;
    const index = bounties.findIndex((bounty) => bounty.guid === guid);
    if (index >= 0) {
      const bounty = bounties[index];
      const a = {
        author: assertion.author,
        bid: assertion.bid,
        metadata: assertion.metadata,
        verdicts: assertion.verdicts,
      };
      bounty.assertions.push(a);
      const modified = bounty.artifacts.map((file, index) => {
        const f = file;
        if (!a.verdicts[index]) {
          f.good++;
        }
        f.total++;
        file.assertions.push({
          author: assertion.author,
          bid: assertion.bid,
          verdict: a.verdicts[index],
          metadata: assertion.metadata
        });
        return f;
      });
      bounty.artifacts = modified;
      bounty.update = true;
      bounties[index] = bounty;
    }
    this.setState({bounties: bounties});
  }

  getData() {
    const http = this.http;
    const uuid = Uuid();
    this.addRequest(strings.requestAllData, uuid);
    http.listenForAssertions(this.updateOnAssertion);
    const bounties = this.state.bounties.slice();
    const promises = bounties.map((bounty) => {
      return http.getBounty(bounty)
        .then(b => {
          if (b == null) {
            return bounty;
          }
          b.updated = bounty.updated;
          if (JSON.stringify(b) !== JSON.stringify(bounty) || bounty.updated) {
            b.updated = true;
          }
          return b;
        });
    });
    return Promise.all(promises).then((values) => {
      // get updated state after download finishes
      const bounties = this.state.bounties.slice();
      values.forEach((value) => {
        const foundIndex = bounties.findIndex((bounty) => bounty.guid === value.guid);
        if (foundIndex >= 0) {
          bounties[foundIndex] = value;
        }
      });
      this.setState({bounties: bounties});
      this.removeRequest(strings.requestAllData, uuid);
    });
  }

  getPropsForChild() {
    const {host: url} = config;
    const { state: { active, bounties, walletList, requestsInProgress, address } } = this;
    return({
      url,
      active,
      address,
      bounties,
      walletList,
      requestsInProgress,
      onError: this.onPostError,
      addRequest: this.addRequest,
      removeRequest: this.removeRequest,
      onCreateOffer: this.onCreateOffer,
      onCreateBounty: this.onCreateBounty,
    });
  }

  getWallets() {
    const http = this.http;
    return http.getWallets()
      .then(addresses => addresses.map(address => {
        return({address: address});
      }))
      .then(wallets => {
        const promises = wallets.map((wallet) => {
          const e = http.getEth(wallet.address).then(balance =>
            new BigNumber(balance).dividedBy(new BigNumber(1000000000000000000))
          )
            .then((b) => `${b.toNumber()}`);

          const n = http.getNct(wallet.address).then(balance =>
            new BigNumber(balance).dividedBy(new BigNumber(1000000000000000000))
          )
            .then((b) => `${b.toNumber()}`);

          const promises = [e, n];
          return Promise.all(promises).then(values => {
            return({
              address: wallet.address,
              eth: values[0],
              nct: values[1]
            });
          });
        });
        return Promise.all(promises);
      })
      .then((wallets) => new Promise(resolve =>
        this.setState({walletList: wallets}, resolve)
      ))
      .then(() => http.getUnlockedWallet())
      .then(address => {
        const {state: {walletList}} = this;
        if (!address) {
          this.setState({ address: 0 });
        } else {
          const index = walletList.findIndex((account => account.address === address));
          this.setState({ address: index });
        }
      });
  }

  storeBounties(bounties) {
    if (this.hasLocalStorage()) {
      localStorage.setItem('bounties', JSON.stringify(bounties));
    }
  }

  hasLocalStorage() {
    try {
      localStorage.setItem('x', 'y');
      localStorage.removeItem('x');
      return true;
    } catch(e) {
      return false;
    }
  }

  markSeen() {
    if (this.hasLocalStorage()) {
      localStorage.setItem('seen', JSON.stringify(true));
    }
  }

  preloadLocalStorage() {
    if (this.hasLocalStorage) {
      const bounties = JSON.parse(localStorage.getItem('bounties')) || [];
      const first = !JSON.parse(localStorage.getItem('seen'));
      return {bounties, first};
    } else {
      return {bounties: [], first: true};
    }
  }
}

export default App;
