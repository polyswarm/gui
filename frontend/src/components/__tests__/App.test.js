import React from 'react';
import {shallow, render, mount} from 'enzyme';
import {renderToJson, shallowToJson} from 'enzyme-to-json';
import App from '../App';

it('renders without crashing', () => {
  const wrapper = render(<App />);
  expect(renderToJson(wrapper)).toMatchSnapshot();
});