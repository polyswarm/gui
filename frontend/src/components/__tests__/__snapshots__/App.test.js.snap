// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  class="App"
>
  <div
    class="Landing"
  >
    <header
      class="Header"
    >
      <h2>
        PolySwarm Relay
      </h2>
    </header>
    <div
      class="Landing-Content"
    >
      <h2>
        0x6322c188b4b353acb89c0db66672f0bac421c9e8
      </h2>
      <div
        class="Landing-Chain"
      >
        <div
          class="ChainInfo"
        >
          <div
            class="ChainInfo-Title"
          >
            <h2>
              Before Transfer
            </h2>
          </div>
          <div
            class="ChainInfo-Content"
          >
            <p>
              
            </p>
            <ul>
              <li>
                <b>
                  Mainnet
                </b>
                : 10
              </li>
              <li>
                <b>
                  PolySwarm Sidechain
                </b>
                : 10
              </li>
            </ul>
          </div>
        </div>
        <div
          class="ChainInfo"
        >
          <div
            class="ChainInfo-Title"
          >
            <h2>
              After Transfer
            </h2>
          </div>
          <div
            class="ChainInfo-Content"
          >
            <p>
              
            </p>
            <ul>
              <li>
                <b>
                  Mainnet
                </b>
                : 10
              </li>
              <li>
                <b>
                  PolySwarm Sidechain
                </b>
                : 10
              </li>
            </ul>
          </div>
        </div>
      </div>
      <div
        class="Landing-Nectar"
      >
        <div
          class="NectarField"
        >
          <label
            class="NectarField-Label"
            for="nectar"
          >
            Amount of Nectar
          </label>
          <input
            class="NectarField-Input"
            id="nectar"
            type="number"
            value=""
          >
            
          </input>
        </div>
        <button
          class="Button"
        >
          Transfer!
        </button>
      </div>
    </div>
  </div>
</div>
`;
